nextcloud/vue-richtext
# Implement reference picker

refs https://github.com/nextcloud/server/issues/31667
requires https://github.com/nextcloud/server/pull/35557

* [x] implement reference picker
    * [x] first step: reference provider picker (multi select)
    * [x] second step: vanilla input to enter a link
    * [x] alternative second step if provider is searchable: multi select to search with all search providers supported by the reference provider
* [x] add mechanism to allow apps to inject a custom picker component

The reference picker gets the reference provider list via an initial state provided by the server. Whatever page using the reference picker should have dispatched the `RenderReferenceEvent` to trigger the production of the initial state.

Each reference provider can declare a list of supported unified search providers. Those search provider will be used if no custom link picker is provided. The search UI is pretty similar to the core unified search: small thumbnail, a title and a sub line.

If the search UI does not fit the needs of the reference provider, it is possible to register a custom component to let users choose an element to get a link in the end. This component must emit the `submit` event with the link as event data. It an also emit the `cancel` event to go back to the reference provider picker.

To test this PR, one can:
* use https://github.com/nextcloud/server/pull/35557
* use the `reference-discovery` branch of the integration_github app (don't forget to link with `@nextcloud/vue-richtext` while having the `enh/noid/reference-picker` branch checked out)
* go to the "Connected accounts" admin settings where a test textarea+button can be found to use the reference picker

Remaining questions/issues/todos:
* There is no translations in vue-richtext but this brings strings that should be translated.
* We should find a way to use this reference picker in the `RichText` component (when typing a special character and/or when receiving an event via the NC event bus and/or via a component method call)
